exporter/awsemfexporter/metadata/metadata.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metadata
+
+import (
+	"bufio"
+	"io"
+	"os"
+
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+)
+
+const (
+	containerIDLength = 64
+	defaultCGroupPath = "/proc/self/cgroup"
+)
+
+type Metadata interface {
+	GetHostIdentifier() (string, error)
+	GetEC2InstanceID() (string, error)
+}
+
+type metadata struct {
+	ec2    *ec2metadata.EC2Metadata
+	docker *DockerHelper
+	hostID string
+}
+
+func NewMetadata(s *session.Session, cGroupPath string) Metadata {
+	if cGroupPath == "" {
+		cGroupPath = defaultCGroupPath
+	}
+	return &metadata{
+		ec2: ec2metadata.New(s),
+		docker: &DockerHelper{
+			cGroupPath: cGroupPath,
+		},
+	}
+}
+
+func (m *metadata) isOnEC2() bool {
+	return m.ec2.Available()
+}
+
+func (m *metadata) GetEC2InstanceID() (string, error) {
+	instance, err := m.ec2.GetInstanceIdentityDocument()
+	if err != nil {
+		return "", err
+	}
+	return instance.InstanceID, nil
+}
+
+func (m *metadata) getContainerID() (string, error) {
+	return m.docker.getContainerID()
+}
+
+func (m *metadata) GetHostIdentifier() (string, error) {
+	if m.hostID != "" {
+		return m.hostID, nil
+	}
+	var id string
+	var err error
+	if m.isOnEC2() {
+		id, err = m.GetEC2InstanceID()
```

Comment:
Shouldn't we check container ID before EC2 host ID? If multiple collectors are in containers on the same EC2 node, they'll all have the same host identifier, which I believe is what you're trying to avoid.

Reply:
But do consider my suggestion of UUID to not have to think about such tricky conditions :)

Reply:
I will move ContainerID check before EC2. Regarding the UUID, I think the CollectorID is more readable from customer's perspective. For example, they can know which EC2 instance is emitting that log stream.